### PR TITLE
feat: Qwen-Image-Edit inference and training

### PR DESCRIPTION
**Pull request by kohya-ss**
- Use Image Dataset with Control Images: https://github.com/kohya-ss/musubi-tuner/blob/main/src/musubi_tuner/dataset/dataset_config.md#sample-for-image-dataset-with-control-images 
    - `flux_kontext_no_resize_control` is not supported. Control images are resized/cropped to the target image size.
- Use Qwen-Image-Edit weights from [official HF page](https://huggingface.co/Qwen/Qwen-Image-Edit/tree/main/transformer) (specify `00001` to `--dit`) or [ComyUI repackaged](https://huggingface.co/Comfy-Org/Qwen-Image-Edit_ComfyUI/tree/main/split_files/diffusion_models) (use bf16, not tested.)
- Latent and Text Encoder output caching commands are same as Qwen-Image. However re-run both to include control images.
- Add `--edit` option to Text Encoder output caching, inference and training script.
- Specify `--control_image_path` to specify control image for inference script like [FLUX.1 Kontext](https://github.com/kohya-ss/musubi-tuner/blob/main/docs/flux_kontext.md).
- The sample prompt (during training) must include `--ci path/to/control.png` (or jpg etc.) option to specify the control image.

### CODE REVIEW

**Comment by kohya-ss** [comment]
Token max length calculation is fixed. Please re-run Text Encoder output caching.

**Comment by mliand** [comment]
image
it seems that there are some issues with the patch processing part.

**Comment by sdbds** [comment]
image
After testing, normal training works, but it seems that under the bucket [1328,1328], 50GB VRAM is required

**Comment by kohya-ss** [comment]
> After testing, normal training works, but it seems that under the bucket [1328,1328], 50GB VRAM is required

Thank you for testing! 

The model weights consume just under 40GB, and the tensor sequence length is twice that of Qwen-Image training, so memory consumption is quite high. Although there are quality issues, specifying `--fp8_base` and `--fp8_scaled` may be more practical.

**Comment by kohya-ss** [comment]
By removing unnecessary variables, the peak memory was reduced by about 1GB when training 1328x1328 Qwen-Image-Edit.

**Comment by kohya-ss** [comment]
The original Diffusers implementation required the generated image and the control image to have the same resolution, but this appears to have been a bug in Diffusers.

https://github.com/huggingface/diffusers/pull/12188
https://github.com/huggingface/diffusers/pull/12190

Allowing control images of any size, similar to training FLUX.1 Kontext, may reduce memory consumption.

**Comment by sdbds** [comment]
I feel that qwen edit learns very slowly, and I'm not sure if it's a problem with the model itself...

**Comment by sdbds** [comment]
image

https://github.com/modelscope/DiffSynth-Studio/pull/814

I checked the diffsyth code, and it seems they modified the TE template  and drop idx when use qwen-image edit

**Comment by kohya-ss** [comment]
> [modelscope/DiffSynth-Studio#814](https://github.com/modelscope/DiffSynth-Studio/pull/814)
> 
> I checked the diffsyth code, and it seems they modified the TE template and drop idx when use qwen-image edit

We are already doing with that 😄 : https://github.com/kohya-ss/musubi-tuner/blob/d33969cf0701041fe682b1a2f4e6a13f01af3eed/src/musubi_tuner/qwen_image/qwen_image_utils.py#L375

**Comment by mliand** [comment]
some difficulties in the same training.

**Comment by kohya-ss** [comment]
`dataset_config.md` now has following two new options for Qwen-Image-Edit.

- `qwen_image_edit_no_resize_control` : This is the same as FLUX.1 Kontext, does not resize the control image.
- `qwen_image_edit_control_resolution`: Resizes to a specific bucket size. To resize the control images the same as the official code, specify [1024,1024].

If neither option is present, the control image will be resized to the same size as the target image. Please re-run latent and Text Encoder caching if these options are changed.

`qwen_image_generate_image.py` has following two new options.

- `--resize_control_to_image_size`: Resize the control image for generation size.
- `--resize_control_to_official_size`: Resize the control image to match official size (1M pixels keeping aspect ratio).

FLUX.1 Kontext caching/training are also updated, so the cache must be re-cached.

File: src/musubi_tuner/qwen_image_generate_image.py [hunk 1]
```diff
@@ -573,32 +657,41 @@ def generate(
...
                     timestep=timestep / 1000,
                     guidance=guidance,
                     encoder_hidden_states_mask=mask,
                     encoder_hidden_states=embed,
                     img_shapes=img_shapes,
                     txt_seq_lens=txt_seq_lens,
-                )[0]
+                )
+                if is_edit:
+                    noise_pred = noise_pred[:, : latents.shape[1], :]  # trim to latents shape
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply src/musubi_tuner/qwen_image_generate_image.py:679]: `image_seq_len` is for calculating shift value, so `latents.shape[1]` is more intuitive here, because of `latent_model_input = torch.cat([latents, control_latent], dim=1)`.

File: src/musubi_tuner/qwen_image_generate_image.py [hunk 2]
```diff
@@ -573,32 +657,41 @@ def generate(
...
                         timestep=timestep / 1000,
                         guidance=guidance,
                         encoder_hidden_states_mask=negative_mask,
                         encoder_hidden_states=negative_embed,
                         img_shapes=img_shapes,
                         txt_seq_lens=negative_txt_seq_lens,
-                    )[0]
+                    )
+                if is_edit:
+                    neg_noise_pred = neg_noise_pred[:, : latents.shape[1], :]  # trim to latents shape
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply src/musubi_tuner/qwen_image_generate_image.py:693]: Same as above.

File: src/musubi_tuner/qwen_image_generate_image.py [hunk 3]
```diff
@@ -347,6 +361,38 @@ def load_dit_model(
 # endregion
 
 
+def prepare_image_inputs(
+    args: argparse.Namespace, device: torch.device, vae: AutoencoderKLQwenImage
+) -> tuple[torch.Tensor, np.ndarray]:
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply src/musubi_tuner/qwen_image_generate_image.py]: This makes sense.

File: src/musubi_tuner/qwen_image_generate_image.py [hunk 4]
```diff
@@ -805,25 +901,43 @@ def process_batch_prompts(prompts_data: List[Dict], args: argparse.Namespace) ->
...
         "conds_cache": conds_cache_batch,
     }
 
+    if args.edit:
+        vae_for_batch.to(device)  # Move VAE to device for control image encoding
+
+        for i, prompt_args_item in enumerate(all_prompt_args_list):
+            logger.info(f"Preprocessing control image for prompt {i+1}/{len(all_prompt_args_list)}: {prompt_args_item.prompt}")
+            assert prompt_args_item.control_image_path is not None, "Qwen-Image-Edit requires control_image_path"
+            control_data = prepare_image_inputs(args, device)
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply src/musubi_tuner/qwen_image_generate_image.py]: sure!

**Comment by kohya-ss** [comment]
Documents are added.

**Comment by sdbds** [comment]
After testing, inputting [1024,1024] seems to result in significantly better training performance. Perhaps it needs to match the official input of 1M pixels...
Also, the default learning rate of 1E-4 seems a bit high for this model? It feels like it's easily overfitting.

**Comment by kohya-ss** [comment]
> After testing, inputting [1024,1024] seems to result in significantly better training performance. Perhaps it needs to match the official input of 1M pixels...

By specifying `qwen_image_edit_control_resolution = [1024, 1024]` in the dataset config, The control images are fixed to 1M pixels regardless of the resolution of the training images.

However, it seems like it would also be a good idea to set the training images to [1024,1024].

> Also, the default learning rate of 1E-4 seems a bit high for this model? It feels like it's easily overfitting.

Thank you! I think the default should be a conservative value, and I also use a lower learning rate than other models, so I changed it to 5e-5. I also changed rank(dim) to 16 because LoRA becomes huge with 32.

**Comment by sdbds** [comment]
@kohya-ss 
Because the effect is very strange, I specifically asked the official technical staff, and they told me that they use specialized bucket training for the edit model.
I think we can consider setting up these specialized buckets, like wan 2.1.
```
1024, 1024
1168, 864
864, 1168
1280, 720
720, 1280
896, 1120
1120, 896
800, 1200
1200, 800
```

**Comment by kohya-ss** [comment]
> Because the effect is very strange, I specifically asked the official technical staff, and they told me that they use specialized bucket training for the edit model.
> I think we can consider setting up these specialized buckets, like wan 2.1.

Thank you! Hmm... It would probably be better to make this available as an option. 

It might be a good idea to add the `use_qwen_image_default_buckets` option in the dataset config, or to allow users to specify any resolution using the `custom_bucket_resolutions` option, etc.

**Comment by BenDes21** [comment]
Hi, --resize_control_to_official_size + --control_image_path + --prompt seems to be not recognized as arguments, getting **qwen_image_train_network.py: error: unrecognized arguments**
